Scripts/Sensors/ROS/RaycastLidarPointCloud2ROS.cs

```diff
@@ -89,9 +93,98 @@ private PointCloud2 GeneratePointCloud2(NativeArray<Vector3> points)
             return pointCloud;
         }
 
+        private PointCloud2 GeneratePointCloud2(NativeArray<Vector3> points, NativeArray<LidarReading> readings)
+        {
+            PointCloud2 pointCloud = new PointCloud2();
+            pointCloud.Header = new Std.Header()
+            {
+                FrameId = sensor.frameId,
+                Timestamp = TimeHandler.Instance.TimeDouble
+            };
+            pointCloud.Fields.AddRange(
+                new List<PointField>()
+                {
+                    new PointField()
+                    {
+                        Name = "x",
+                        Offset = 0,
+                        Datatype = PointField.Types.DataType.Float32 + 1,
+                        Count = 1
+                    },
+                    new PointField()
+                    {
+                        Name = "z",
+                        Offset = 4,
+                        Datatype = PointField.Types.DataType.Float32 + 1,
+                        Count = 1
+                    },
+                    new PointField()
+                    {
+                        Name = "y",
+                        Offset = 8,
+                        Datatype = PointField.Types.DataType.Float32 + 1,
+                        Count = 1
+                    },
+                    new PointField()
+                    {
+                        Name = "intensity",
+                        Offset = 12,
+                        Datatype = PointField.Types.DataType.Int32 + 1,
+                        Count = 1
+                    },
+                    new PointField()
+                    {
+                        Name = "ring",
+                        Offset = 16,
+                        Datatype = PointField.Types.DataType.Int32 + 1,
+                        Count = 1
+                    },
+                    new PointField()
+                    {
+                        Name = "time",
+                        Offset = 20,
+                        Datatype = PointField.Types.DataType.Uint32 + 1,
+                        Count = 1
+                    }
+                }
+            );
+            var numOfPoints = readings.Count(r => r.IsValid);
+            var pointSize = sizeof(float) * 3 + sizeof(int)*3;
+            var byteLength = numOfPoints * pointSize;
+            pointCloud.Height = 1;
+            pointCloud.Width = (uint) numOfPoints;
+            pointCloud.IsBigEndian = false;
+            pointCloud.PointStep = (uint) pointSize;
+            pointCloud.IsDense = true;
+            byte[] bytes = new byte[byteLength];
+            int j = 0;
+            for(int i = 0; i < points.Length; i++)
+            {
+                if (!readings[i].IsValid) continue;
+
+                Buffer.BlockCopy( BitConverter.GetBytes( points[i].x ), 0, bytes, j*pointSize, 4 );
```

Comment:
instead of j*pointSize for every line, instantiate a variable to keep track of this. Inside for loop over points, this is unnecessary computation..

The best way would be to use summation instead of multiplication. Init var to 0 and every loop add the pointSize to it

Reply:
In the future, NativeArray<Vector3> points could be made generic, and not Vector3. Then, based on configuration it can our defined struct LidarPoints (only Vec3), LidarPointsWithIntensity, LidarPointsWithIntensityAndRing, etc.. Then you could again just use block copy. 

This change impacts whole lidar and sonar pipeline with RaycastJobHelper